RecoHI/HiJetAlgos/plugins/HiPuRhoProducer.cc

```diff
@@ -0,0 +1,645 @@
+  for (std::size_t i = 0; i < size; ++i) {
+    order.emplace_back(std::make_pair(i, etaEdgeLow_[i]));
+  }
+
+  std::sort(
+      order.begin(), order.end(), [](auto const& pair0, auto const& pair1) { return pair0.second < pair1.second; });
+
+  std::vector<double> sortedEtaEdgeLow(size);
+  std::vector<double> sortedEtaEdgeHigh(size);
+
+  auto mapToRhoOut = std::make_unique<std::vector<double>>(size);
+  auto mapToRhoExtraOut = std::make_unique<std::vector<double>>(size);
+  auto mapToRhoMOut = std::make_unique<std::vector<double>>(size);
+  auto mapToNTowOut = std::make_unique<std::vector<int>>(size);
+
+  for (std::size_t i = 0; i < size; ++i) {
+    auto const& index = order[i].first;
+
+    sortedEtaEdgeLow[i] = etaEdgeLow_[index];
+    sortedEtaEdgeHigh[i] = etaEdgeHi_[index];
+
+    (*mapToRhoOut)[i] = rho_[index];
+    (*mapToRhoExtraOut)[i] = rhoExtra_[index];
+    (*mapToRhoMOut)[i] = rhoM_[index];
+    (*mapToNTowOut)[i] = nTow_[index];
+  }
+
+  auto mapToRhoMedianOut = std::make_unique<std::vector<double>>(size);
+
+  for (uint32_t index = medianWindowWidth_; index < size - medianWindowWidth_; ++index) {
+    auto centre = mapToRhoOut->begin() + index;
+    std::vector<float> rhoRange(centre - medianWindowWidth_, centre + medianWindowWidth_);
+    std::nth_element(rhoRange.begin(), rhoRange.begin() + medianWindowWidth_, rhoRange.end());
+    (*mapToRhoMedianOut)[index] = rhoRange[medianWindowWidth_];
+  }
+
+  auto mapEtaRangesOut = std::make_unique<std::vector<double>>();
+
+  mapEtaRangesOut->assign(sortedEtaEdgeLow.begin(), sortedEtaEdgeLow.end());
+  mapEtaRangesOut->push_back(sortedEtaEdgeHigh.back());
+
+  auto mapToTowExcludePtOut = std::make_unique<std::vector<double>>(std::move(towExcludePt_));
+  auto mapToTowExcludePhiOut = std::make_unique<std::vector<double>>(std::move(towExcludePhi_));
+  auto mapToTowExcludeEtaOut = std::make_unique<std::vector<double>>(std::move(towExcludeEta_));
+
+  iEvent.put(std::move(mapEtaRangesOut), "mapEtaEdges");
+  iEvent.put(std::move(mapToRhoOut), "mapToRho");
+  iEvent.put(std::move(mapToRhoMedianOut), "mapToRhoMedian");
+  iEvent.put(std::move(mapToRhoExtraOut), "mapToRhoExtra");
+  iEvent.put(std::move(mapToRhoMOut), "mapToRhoM");
+  iEvent.put(std::move(mapToNTowOut), "mapToNTow");
+  iEvent.put(std::move(mapToTowExcludePtOut), "mapToTowExcludePt");
+  iEvent.put(std::move(mapToTowExcludePhiOut), "mapToTowExcludePhi");
+  iEvent.put(std::move(mapToTowExcludeEtaOut), "mapToTowExcludeEta");
+}
+
+// ------------ method fills 'descriptions' with the allowed parameters for the module  ------------
+void HiPuRhoProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  //The following says we do not know what parameters are allowed so do no validation
+  // Please change this to state exactly what you do use, even if it is no parameters
+  edm::ParameterSetDescription desc;
+  desc.setUnknown();
+  descriptions.addDefault(desc);
```

Comment:
this should implement all parameters properly, after that the explicit RecoHI/HiJetAlgos/python/hiPuRhoProducer_cfi.py file can be removed from the source code